src/nti/site/site.py

```diff
@@ -314,51 +315,108 @@ def _btlar_after_setstate(self):
     # these ``_subscribers`` is to implement ``getAllUtilitiesRegisteredFor``; that's rarely called,
     # right? Maybe rare enough that we could implement a different mechanism that doesn't need to
     # persistently store the whole list at all.
+    #
+    # IMPORTANT: Once a parent mapping has become a BTree, all of its children mappings
+    # must ALSO become BTrees. This is because the mappings are mutated in-place as registrations
+    # are added or removed, and doing that doesn't let the BTree know that it needs to become
+    # ``_p_changed``; trying to figure out which child dictionary actually changed *could* be done
+    # (err, somehow) but we'd actually need to set the ``_p_changed`` on a BTree *bucket*,
+    # and those aren't directly accessible. So: once a BTree, it's BTrees all the way down.
+
+    def __mapping_needs_converted(self, path, mapping, parent_is_btree, btree_type):
+        # We need to convert if the parent is a btree (and we're not a btree already)
+        # OR we're not a btree and we're too long
+        regs_is_btree = isinstance(mapping, btree_type)
+
+        needs_converted = (
+            parent_is_btree and not regs_is_btree
+        ) or (
+            not regs_is_btree and len(mapping) > self.btree_map_threshold
+        )
+        if needs_converted:
+            logger.info("Converting bucket (k=%s, len=%d) to %s.",
+                        path, len(mapping), btree_type)
+
+        return needs_converted
+
+    def __btree_map(self, path, mapping, depth, max_depth,
+                    parent_is_btree,
+                    btree_type):
+        # type: (tuple, dict, int, int, bool, type) -> dict
+        #
+        # path is just for information
+        if self.__mapping_needs_converted(
+                path, mapping, parent_is_btree, btree_type
+        ):
+            mapping = btree_type(mapping)
+            # Force conversion of child leaves
+            parent_is_btree = True
+            #changed_below = True
+        else:
+            parent_is_btree = parent_is_btree or isinstance(mapping, btree_type)
+
+        replacement_vals = {}
+        if depth < max_depth:
+            # Still have a layer of dicts to go through
+            for k, v in mapping.items():
+                new_v = self.__btree_map(
+                    path + (k,),
+                    v,
+                    depth + 1,
+                    max_depth,
+                    parent_is_btree,
+                    btree_type
+                )
+                if new_v is not v:
```

Comment:
Is `is not` the correct check here? Is that checking if the instance is the same?

Reply:
Correct. If we mutated something in place, there's no need to keep holding on to it here.

I had a more complicated version of this that tracked if anything anywhere beneath something changed and rewrote the tree all the way up, but that was complex and turned out to be unnecessary if I started conversion from the bottom but tracked if a parent was (or became) a btree. The `parent_is_btree` argument simplified everything a bunch.